vSphere Integration

qduk opened the pull request:
<!--
    Thank you for your interest in contributing to Single Source of Truth! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->

# Closes: #500 

## What's Changed

<!--
    Please include:
    - A summary of the proposed changes
    - A sectioned breakdown for larger features under ## subheadings
    - Screenshots, example payloads where relevant:
      - Before/After for bugfixes
      - Using a new feature
-->

Adds the ability to sync directly with a vCenter instance. Currently, this is implemented as a one-way sync from vCenter into Nautobot. The following models are synced from vCenter:

- DataCenter -> ClusterGroup
- Cluster -> Cluster
- Virtual Machine -> Virtual Machine
- VM Interface -> VM Interface 
- IP Addresses -> IP Addresses 

<!--
![Screenshot 2025-02-18 at 12 31 42 PM](https://github.com/user-attachments/assets/6aa934d0-178e-4559-a948-69152d0e960c)

For configuration, this uses the same construct of a `Config Instance` like the Infoblox integration.
![Screenshot 2025-02-18 at 12 32 21 PM](https://github.com/user-attachments/assets/aaaeaf66-9d6b-4bff-9a7f-7921724a5c82)

## To Do

    Please feel free to update todos to keep track of your own notes for WIP PRs.
-->
- [X] Explanation of Change(s)
- [X] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [X] Attached Screenshots, Payload Example
- [X] Unit, Integration Tests
- [X] Documentation Updates (when adding/changing features)
- [X] Add screenshots from Job Results to Documentation
- [X] Test against live vSphere with various setting configurations
- [X] Add @h4ndzdatm0ld as co-contributor

qduk commented:
@jdrew82 Do you know why the migration test is failing? I would think that dependent migration exists. I'm not sure if I should just remove that one as a dependency or not.

jdrew82 reviewed (changes requested):
Few things that I've noticed.

qduk commented on nautobot_ssot/integrations/vsphere/diffsync/adapters/adapter_nautobot.py:
```diff
@@ -0,0 +1,57 @@
+)
+
+
+class Adapter(NautobotAdapter):
```
Updated to `NBAdapter`

qduk commented on nautobot_ssot/integrations/vsphere/diffsync/adapters/__init__.py:
```diff
@@ -0,0 +1,6 @@
+"""Collection of adapters."""
+
+from .adapter_nautobot import Adapter
```
Updated to `NBAdapter`

qduk commented on development/nautobot_config.py:
```diff
@@ -267,6 +268,9 @@
         "servicenow_instance": os.getenv("SERVICENOW_INSTANCE", ""),
         "servicenow_password": os.getenv("SERVICENOW_PASSWORD", ""),
         "servicenow_username": os.getenv("SERVICENOW_USERNAME", ""),
+        "vsphere_username": os.getenv("VSPHERE_USERNAME", ""),
```
It totally is. I originally wasn't so I had these here. Thanks for catching that. Removed!

qduk commented on nautobot_ssot/tests/vsphere/test_vsphere_diffsync_models.py:
```diff
@@ -0,0 +1,830 @@
+            config=self.config,
+            cluster_filters=None,
+        )
+        # VirtualMachine.objects.all().delete()
```
Yes.

qduk commented on nautobot_ssot/tests/vsphere/test_vsphere_diffsync_models.py:
```diff
@@ -0,0 +1,830 @@
+            config=self.config,
+            cluster_filters=None,
+        )
+        # VirtualMachine.objects.all().delete()
```
Double yes.

qduk commented on nautobot_ssot/contrib/adapter.py:
```diff
@@ -116,7 +116,10 @@ def _handle_single_parameter(self, parameters, parameter_name, database_object,
         if custom_relationship_annotation:
             parameters[parameter_name] = self._handle_custom_relationship_to_many_relationship(
-                database_object, diffsync_model, parameter_name, custom_relationship_annotation
+                database_object,
```
I don't think so. It is passing the ruff test.

qduk replied:
You are totally right. I had VSCode format upon save using black. I guess I never changed it to ruff! I went and reformatted.

jdrew82 commented on tasks.py:
```diff
@@ -51,7 +51,7 @@ def is_truthy(arg):
     {
         "nautobot_ssot": {
-            "nautobot_ver": "2.3.1",
+            "nautobot_ver": "2.3.16",
```
I don't think we want to be doing this.

qduk replied:
Yeah...I'd agree. Not too sure why I initially changed this. Reverted!

jdrew82 reviewed (changes requested):
Few things

qduk commented on nautobot_ssot/integrations/vsphere/defaults.py:
```diff
@@ -0,0 +1,8 @@
+"""Default settings to use across different files."""
```
All configuration options have been moved tot eh Config model with defaults put into place. I agree that it makes way more sense.

jdrew82 commented on docs/admin/integrations/vsphere_setup.md:
```diff
@@ -0,0 +1,122 @@
+
+The default value says that the status of `POWERED_OFF` translates to the `Offline` status in Nautobot, `POWERED_ON` to the `Active` status in Nautobot and `SUSPENDED` to the `Suspended` status in Nautobot.
+
+> Note: Installation of the vSphere integration with ensure that the above 3 statuses exist. If you change this setting, you will have to ensure that the Nautobot status you are mapping to has already been created. It is highly recommended to leave the default settings here.
```
```suggestion
> Note: Installation of the vSphere integration will ensure that the above 3 Statuses exist. If you change this setting, you will have to ensure that the Nautobot Status you are mapping to has already been created. It is highly recommended to leave the default settings here.
```

jdrew82 commented on docs/admin/integrations/vsphere_setup.md:
```diff
@@ -0,0 +1,122 @@
+
+### Configuring Virtual Machine Status Map
+
+Virtual Machine Status Map is a mandatory setting used to map the status of a machine in vSphere to a status in Nautobot.
```
```suggestion
Virtual Machine Status Map is a mandatory setting used to map the status of a machine in vSphere to a Status in Nautobot.
```

jdrew82 replied:
Status is a proper name of an object in Nautobot so we should always capitalize it.

jdrew82 commented on docs/admin/integrations/vsphere_setup.md:
```diff
@@ -0,0 +1,122 @@
+
+### Configuring Virtual Machine IP Status Map
+
+Virtual Machine IP Status Map is a mandatory setting used to map the status of and IP address in vSphere to a status in Nautobot.
```
```suggestion
Virtual Machine IP Status Map is a mandatory setting used to map the status of and IP address in vSphere to a Status in Nautobot.
```

jdrew82 commented on docs/admin/integrations/vsphere_setup.md:
```diff
@@ -0,0 +1,122 @@
+
+The default maps the `PREFERRED` and `UNKNOWN` states from vSphere to the `Active` and `Reserved`, respectively. 
+
+> Note: Installation of the vSphere integration with ensure that the above 2 statuses exist. If you change this setting, you will have to ensure that the Nautobot status you are mapping to has already been created. It is highly recommended to leave the default settings here.
```
```suggestion
> Note: Installation of the vSphere integration will ensure that the above 2 Statuses exist. If you change this setting, you will have to ensure that the Nautobot Status you are mapping to has already been created. It is highly recommended to leave the default settings here.
```

jdrew82 commented on docs/admin/integrations/vsphere_setup.md:
```diff
@@ -0,0 +1,122 @@
+
+### Configuring Virtual Machine Interface Status Map
+
+Virtual Machine Interface Status Map is a mandatory setting used to map the status of an Interface in vSphere to a status in Nautobot.
```
```suggestion
Virtual Machine Interface Status Map is a mandatory setting used to map the status of an Interface in vSphere to a Status in Nautobot.
```

jdrew82 commented on docs/admin/integrations/vsphere_setup.md:
````diff
@@ -0,0 +1,122 @@
+    "NOT_CONNECTED": false
+}
+```
+> Note: Installation of the vSphere integration with ensure that the above 2 statuses exist. If you change this setting, you will have to ensure that the Nautobot status you are mapping to has already been created. It is highly recommended to leave the default settings here.
````
```suggestion
> Note: Installation of the vSphere integration will ensure that the above 2 Statuses exist. If you change this setting, you will have to ensure that the Nautobot Status you are mapping to has already been created. It is highly recommended to leave the default settings here.
```

jdrew82 commented on docs/admin/integrations/vsphere_setup.md:
```diff
@@ -0,0 +1,122 @@
+> Note: Installation of the vSphere integration with ensure that the above 2 statuses exist. If you change this setting, you will have to ensure that the Nautobot status you are mapping to has already been created. It is highly recommended to leave the default settings here.
+## Custom Fields, Statuses and Tags
+
+The vSphere Integration requires the following Nautobot custom fields and tags to function correctly. These are created automatically when Nautobot is started and care should be taken to ensure these are not deleted. 
```
```suggestion
The vSphere Integration requires the following Nautobot Custom Fields and Tags to function correctly. These are created automatically when Nautobot is started and care should be taken to ensure these are not deleted. 
```

jdrew82 reviewed (changes requested):
Just a few things I noticed.

qduk commented:
> > @jdrew82 Do you know why the migration test is failing? I would think that dependent migration exists. I'm not sure if I should just remove that one as a dependency or not. It does pass locally.
> 
> Getting this error in the Nautobot container when running `invoke start` locally and in the CI:
> 
> > CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0014_alter_sync_start_time_alter_synclogentry_timestamp, 0014_ssotvsphereconfig in nautobot_ssot). To fix them run 'python manage.py makemigrations --merge'
> 
> To fix, I renamed `0014_ssotvsphereconfig.py` to `0015_ssotvsphereconfig.py` and updated the `0015_ssotvsphereconfig.py` dependencies to use 0014 migration:
> 
> Before
> 
> ```python
> class Migration(migrations.Migration):
>     dependencies = [
>         ("nautobot_ssot", "0013_update_solarwinds_customfields"),
>     ]
> ```
> 
> After
> 
> ```python
> class Migration(migrations.Migration):
>     dependencies = [
>         ("nautobot_ssot", "0014_alter_sync_start_time_alter_synclogentry_timestamp"),
>     ]
> ```
> 
> It looks like you probably pulled in newer migrations script from the develop branch and forgot to triage.

@thomasphung You're totally right. Thanks for bringing this to my attention. Fixed!

qduk commented on nautobot_ssot/integrations/vsphere/signals.py:
```diff
@@ -0,0 +1,136 @@
+        defaults={
+            "remote_url": str(config.get("vcenter_url", "https://replace.me.local")),
+            "secrets_group": secrets_group,
+            "verify_ssl": bool(config.get("verify_ssl", False)),
```
I changed both to prepend them with `vsphere`. I feel that made the most sense. Great idea!

qduk commented on nautobot_ssot/integrations/vsphere/diffsync/adapters/adapter_vsphere.py:
```diff
@@ -0,0 +1,313 @@
+
+    def load_cluster_groups(self):
+        """Load Cluster Groups (DataCenters)."""
+        clustergroups = self.client.get_datacenters().json()["value"]
```
Updated to check if the connection attempt authenticated or not. If unsuccessful, throw an error and put a log message. 
![Screenshot 2025-05-30 at 1 51 37 PM](https://github.com/user-attachments/assets/c0695d57-4436-457a-9b23-aca44b0e841c)

thomasphung replied:
Interesting, when I updated the `_authenticate()` function to add the session token, the sync jobs run without errors. I'm looking back at my previously failed runs and I do have an instance of that debug line:

> Loading ClusterGroups {'error_type': 'UNAUTHENTICATED', 'messages': [{'args': [], 'default_message': 'Authentication required.', 'id': 'com.vmware.vapi.endpoint.method.authentication.required'}], 'challenge': 'Basic realm="VAPI endpoint",SIGN realm=438b2f3204259f3c2228cb774c6a68d872f1d633,service="VAPI endpoint",sts="<redacted_uri>/sts/STSService/vsphere.local"'}

thomasphung replied:
Hm, if I'm not going through Nautobot, I'm able to fetch VM data without needing to set the token in the session header. The POST response has a set-cookie header:
> {'date': 'Wed, 04 Jun 2025 17:21:21 GMT', 'set-cookie': 'Path=/rest; Secure; HttpOnly, vmware-api-session-id=9f11080fbb360d286e4585facfeebaf1; Path=/api; Secure; HttpOnly', 'expires': 'Thu, 01 Jan 1970 00:00:00 GMT', 'content-type': 'application/json', 'x-envoy-upstream-service-time': '110', 'server': 'envoy', 'transfer-encoding': 'chunked'}

I guess for some reason the session token isn't stored as cookies on my end?

Snippet: https://github.com/thomasphung/vsphere-api-test/blob/main/vsphere_test.py

thomasphung replied:
I'm getting different response headers between my standalone script and the Nautobot job to the same vCenter instance even though both have session token stored in cookies:

> {'date': 'Wed, 04 Jun 2025 22:09:31 GMT', 'content-type': 'application/json', 'x-envoy-upstream-service-time': '206', 'server': 'envoy', 'transfer-encoding': 'chunked'}

> {'date': 'Wed, 04 Jun 2025 22:06:16 GMT', 'www-authenticate': 'Basic realm="VAPI endpoint",SIGN realm=438b2f3204259f3c2228cb774c6a68d872f1d633,service="VAPI endpoint",sts="<redacted>/sts/STSService/vsphere.local"', 'content-type': 'application/json', 'x-envoy-upstream-service-time': '0', 'server': 'envoy', 'transfer-encoding': 'chunked'}

qduk replied:
Interesting. I wonder what I get back in mine. I haven't had time to check this out today or yesterday but I should have an hour or two to mess with this tomorrow. Looks like I'll need to fix a test as well. I appreciate the comments!

qduk replied:
@thomasphung These are the headers I have in the response before my `update` call. 

```
{'date': 'Fri, 06 Jun 2025 17:40:44 GMT', 'set-cookie': 'vmware-api-session-id=818634afd439f5fa8f2c99852ecef55a; Path=/rest; Secure; HttpOnly, vmware-api-session-id=818634afd439f5fa8f2c99852ecef55a; Path=/api; Secure; HttpOnly', 'expires': 'Thu, 01 Jan 1970 00:00:00 GMT', 'content-type': 'application/json', 'x-envoy-upstream-service-time': '91', 'vary': 'Accept-Encoding', 'content-encoding': 'gzip', 'transfer-encoding': 'chunked'}
```

But mine does work fine with the added line of updating the headers. I'm testing against version 8 of vCenter but I don't really think that would make a difference? Just based on the fact that you are getting something different using standalone versus the Nautobot job. To be safe, I'll include the `update` call for the headers.

I did some slight refactoring in my latest commits. Can you run it again on your end?

thomasphung replied:
Yep all works well!

thomasphung commented on docs/admin/integrations/vsphere_setup.md:
```diff
@@ -0,0 +1,127 @@
+| Default Cluster Name          | vSphere Default Cluster                                                        | Denotes the default Cluster name used in the sync if `Use Clusters` is set to Fals.e.                         |
+| Default Cluster Type          | VMWare vSphere                                                                 | Denotes the default Cluster Type to set in the sync.                                                          |
+
+Each vShere configuration must be linked to an [External integration](https://docs.nautobot.com/projects/core/en/stable/user-guide/platform-functionality/externalintegration/?h=external+int) describing the vSphere instance. The following External Integration fields must be defined for integration to work correctly:
```
vSphere typo

qduk replied:
Addressed. Thank you!

thomasphung commented on docs/user/integrations/vsphere.md:
```diff
@@ -0,0 +1,39 @@
+
+![Dashboard View](../../images/vsphere_dashboard.png)
+
+From the dashboard, you can view more detail about the vSPhere sync by clicking the `VMWare vSphere -> Nautobot` link. This view shows the mapping of vSphere objects to Nautobot objects so you know exactly what will get created in Nautobot when this sync is ran. It will also show you a table view of all the times this sync has been ran.
```
vSphere typo

thomasphung commented on docs/admin/integrations/vsphere_setup.md:
```diff
@@ -0,0 +1,127 @@
+| Sync to Nautobot              | True                                                                           | Allows this config to be used in the job syncing from vSphere to Nautobot                                     |
+| Sync Tagged Only              | True                                                                           | Only take into consideration tagged VMs (on the Nautobot side) when performing the sync.                      |
+| Virtual Machine Status Map    | `{"POWERED_OFF": "Offline", "POWERED_ON": "Active", "SUSPENDED": "Suspended"}` | Maps vSphere Virtual Machine status to Nautobot status.                                                       |
+| Virtual Machine IP Status Map | `{"PREFERRED": "Active", "UNKNOWN": "Reserved"}`                               | Maps vSPhere IP status to Nautobot status                                                                     |
```
vSphere typo

qduk replied:
Addressed. Thank you!

thomasphung commented on nautobot_ssot/integrations/vsphere/api/__init__.py:
```diff
@@ -0,0 +1 @@
+"""REST API module for nautobot_ssot vsphere integration."""
```
vSphere typo

thomasphung commented on nautobot_ssot/integrations/vsphere/api/urls.py:
```diff
@@ -0,0 +1,12 @@
+"""Django urlpatterns declaration for nautobot_ssot vsphere API."""
```
vSphere typo

qduk replied:
Addressed. Thank you!

thomasphung commented on docs/admin/integrations/vsphere_setup.md:
```diff
@@ -0,0 +1,127 @@
+
+Configuration for the vSphere integration is defined in the instance of the `SSOTvSphereConfig` model. Multiple configuration instances are supported. Synchronization jobs take the `Config` parameter which specifies the configuration instance to use. 
+
+To access integration configuration navigate to `Apps -> Installed Apps` and click on the cog icon in the `Single Source of Truth` entry. Then, in the table `SSOT Integration Configs` click on the `vSphere Configuration Instance` link. This will take you to the view where youi can view/modify existing config insances or create new ones.
```
"youi" typo

thomasphung commented on nautobot_ssot/integrations/vsphere/jobs.py:
```diff
@@ -0,0 +1,179 @@
+        query_params={"enable_sync_to_nautobot": True, "job_enabled": True},
+    )
+    cluster_filters = MultiObjectVar(
+        label="Cluster Filters.",
```
No need for period as label

thomasphung commented on nautobot_ssot/integrations/vsphere/jobs.py:
```diff
@@ -0,0 +1,179 @@
+    debug = BooleanVar(description="Enable for more verbose debug logging")
+    config = ObjectVar(
+        model=SSOTvSphereConfig,
+        display_field="SSOT vSphere Config",
```
Not sure why I'm not seeing "SSOT vSphere Config" as the field name. I'm seeing the default "Config" as the field name in the "Sync Now" menu. The image provided in [docs/images/vsphere_job.png](https://github.com/nautobot/nautobot-app-ssot/tree/ab_vsphere_integration/docs/images/vsphere_job.png) also says "Config" so I think this is unintended.

thomasphung reviewed (changes requested):
While I'm here, looking for typos in documentation

jdrew82 commented on nautobot_ssot/integrations/vsphere/diffsync/adapters/adapter_vsphere.py:
```diff
@@ -0,0 +1,312 @@
+    def load_cluster_groups(self):
+        """Load Cluster Groups (DataCenters)."""
+        clustergroups = self.client.get_datacenters().json()["value"]
+        self.job.log_debug(message=f"Loading ClusterGroups {clustergroups}")
```
This needs to be updated to use the new logger pattern.

jdrew82 commented on nautobot_ssot/integrations/vsphere/diffsync/adapters/adapter_vsphere.py:
```diff
@@ -0,0 +1,312 @@
+    def load_virtualmachines(self, cluster, diffsync_cluster):
+        """Load Virtual Machines."""
+        virtual_machines = self.client.get_vms_from_cluster(cluster["cluster"]).json()["value"]
+        self.job.log_debug(message=f"Loading VirtualMachines from Cluster {cluster}: {virtual_machines}")
```
This goes for all instances of the log_debug or log_*.

qduk replied:
All my other non-debug logs should be correct. [Example](https://github.com/nautobot/nautobot-app-ssot/blob/a101761f9c9d8286d504e1136eb6e72a87595ca1/nautobot_ssot/integrations/vsphere/diffsync/adapters/adapter_vsphere.py#L310)

jdrew82 commented on nautobot_ssot/integrations/vsphere/utilities/vsphere_client.py:
```diff
@@ -0,0 +1,169 @@
+    return urllib.parse.urlparse(address)
+
+
+class InvalidUrlScheme(Exception):
```
This is a duplicate of the Exception in the project root folder. Please update to use that one.

jdrew82 reviewed (changes requested):
Just a few things to address.
